### PR TITLE
Layers

### PR DESCRIPTION
**Pull request by sebastienros**
Based on Flows, will have to rebase once Flows is merged.

### CODE REVIEW

**Comment by agriffard** [comment]
I have an idea. Tell if I am right :
If the HtmlField supported tokens (which is not the case yet) and that a {{site.name}} token is created, we could add it in the footer RawHtml Widget?

**Comment by agriffard** [comment]
I don't understand why the 'Global widgets' has a view link displaying the 'about' page and the widgets are repeated twice.

**Comment by agriffard** [comment]
If I add or remove widgets on my layer, most of the time, the pages are not changed (I guess the cache is not refreshed).

**Comment by agriffard** [comment]
Sometimes, I even lose all the widgets in the admin editor after publishing. 
I don't know if it is a problem in the Flow or the Layer.

**Comment by sebastienros** [comment]
Theses are called bugs, thanks, I will check.

**Comment by agriffard** [comment]
Todo : Add Description in Module.txt for Flows and Layers.

**Comment by sebastienros** [comment]
The cache is refreshed only on Publication, can you double-check that it's correctly invalidated? In my tests it worked well.

For the token, I agree, I saw the same thing when doing it. Sure we could keep the copyright portion in the layout.

If you are loosing content on Publish, it's probably that there was an exception during the Update event. I need a repro to find this issue.

**Comment by agriffard** [comment]
I created a content type Project with a TitlePart and Widget stereotype.
On the Page content, I added FlowPart and WidgetsListPart.

I changed the site theme to use TheTheme because with TheBlogTheme, for a the widgets are not displayed because of the Content-Page.cshtml view.

I add a project widget in a Page, in the Content zone of the widgetsList.
When I view it, the title of the project is appearing as the title of the page and vice versa.

I add 2 quotes in WidgetList, center aligned and 50%.
=>The css classes 'widget-align-center widget-size-50' are not rendered.

**Comment by agriffard** [comment]
I add a LinkField on the Project Content type.
I add a widget project to the page, with an url.
=> When I publish the page, the url field is empty in the Edit page (but if I view the page, the url field is well displayed)

**Comment by agriffard** [comment]
I place the WidgetsListPart at the end of the content type.

When it is displayed the widgets added in the Content zone appear before all the parts of the page.

I understand it must be difficult to display the widgets where they are expected. 
At least, can we find a way to display the widgets after the content?

**Comment by agriffard** [comment]
If I have a FlowPart AND a WidgetsListPart, when I add a new widget to the flow, the chrome is not added. Only the content editor of the widget appears.

**Comment by sebastienros** [comment]
@agriffard working on your bug reports now, so far I  created a Project with WidgetsList and Flow and it's fine

image

**Comment by sebastienros** [comment]
I added a LinkField and it works fine too.

**Comment by sebastienros** [comment]
WidgetList doesn't support alignment and size options, not sure how you did it. Maybe something I fixed in the meantime.

**Comment by sebastienros** [comment]
I am working on the recipe right now, pretty sure if you setup a Blog with the current one it will break.

**Comment by sebastienros** [comment]
Blogs recipe updated

**Comment by sebastienros** [comment]
Rebased on master (including BS4A6)

**Comment by agriffard** [comment]
http://localhost:2918/Orchard.Demo/Home/Cache throws an exception :
`RuntimeBinderException: Cannot perform runtime binding on a null reference`

Widget.cshtml, line 4
`var contentTypeClassName = ((string)Model.ContentItem.ContentType).HtmlClassify();`

**Comment by agriffard** [comment]
The alternates views of NavigationItemText are not used so the Menu text with icons are not displayed but I don't understand what has changed.

**Comment by agriffard** [comment]
Can we place the 'Layers' menu below Content after 'Content items'?

**Comment by agriffard** [comment]
I managed to define a Sidebar zone in a custom theme and add a widget inside.

**Comment by agriffard** [comment]
I can delete a layer that have widgets inside. 
Should we prevent this and warn the user?

**Comment by agriffard** [comment]
I enabled Layers on an existing instance (not using the updated recipe) so there were no layers defined.
But in this case, I can add a widget without an associated layer.

**Comment by sebastienros** [comment]
Fixed all comments

**Comment by sebastienros** [comment]
@agriffard request for approval to merge

File: src/Orchard.Cms.Web/Modules/Orchard.Flows/Module.txt [hunk 1]
```diff
@@ -4,7 +4,9 @@ Author: The Orchard Team
...
 Version: 2.0.x
 OrchardVersion: 2.0.x
+Description: Provides a content part allowing users to edit their content based on Widgets.
 Features:
     Orchard.Flows:
         Category: Content
-        Depdendencies: Orchard.Content, Orchard.Widgets
\ No newline at end of file
+        Description: Provides a content part allowing users to edit their content based on Widgets.
+        Dependencies: Orchard.Content, Orchard.Widgets
```
Comment (agriffard) [inline src/Orchard.Cms.Web/Modules/Orchard.Flows/Module.txt]: Orchard.Contents

File: src/Orchard.Cms.Web/Modules/Orchard.Layers/Module.txt [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+Author: The Orchard Team
+Website: http://orchardproject.net
+Version: 2.0.x
+OrchardVersion: 2.0.x
+Description: Enables users to render Widgets across pages of the site based on conditions.
+Features:
+    Orchard.Layers:
+        Category: Content
+        Description: Enables users to render Widgets across pages of the site based on conditions.
+        Dependencies: Orchard.Content, Orchard.Widgets, Orchard.Scripting
```
Comment (agriffard) [inline src/Orchard.Cms.Web/Modules/Orchard.Layers/Module.txt]: Orchard.Contents

File: src/Orchard.Cms.Web/Modules/Orchard.Widgets/Module.txt [hunk 3]
```diff
@@ -4,7 +4,9 @@ Author: The Orchard Team
...
 Version: 2.0.x
 OrchardVersion: 2.0.x
+Description: Provides the default shapes to render Widgets.
 Features:
     Orchard.Widgets:
         Category: Content
-        Depdendencies: Orchard.Content
\ No newline at end of file
+        Description: Provides a part allowing content items to render Widgets in theme zones.
+        Dependencies: Orchard.Content
```
Comment (agriffard) [inline src/Orchard.Cms.Web/Modules/Orchard.Widgets/Module.txt]: Orchard.Contents

File: src/Orchard.Cms.Web/Modules/Orchard.Flows/Module.txt (hunk 1 above)
Comment (agriffard) [inline src/Orchard.Cms.Web/Modules/Orchard.Flows/Module.txt]: Orchard.Contents

**Comment by agriffard** [comment]
In module.txt of Flows, Layers, Widgets:
`Dependencies: Orchard.Contents`
or it will appear as missing dependency.

File: src/Orchard.Cms.Web/Modules/Orchard.Layers/Views/LayerMetadata.Edit.cshtml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+<fieldset class="form-check">
+    <label class="form-check-label">
+        <input asp-for="LayerMetadata.RenderTitle" type="checkbox" class="form-check-input content-preview-select" checked="@Model.LayerMetadata.RenderTitle" /> @T["Render title"]
+        <span class="hint">@T["— If checked, the title will be rendered with the widget."]</span>
+    </label>
+</fieldset>
+
+<div class="row">
+    <fieldset class="form-group col-xs-6">
```
Comment (agriffard) [inline src/Orchard.Cms.Web/Modules/Orchard.Layers/Views/LayerMetadata.Edit.cshtml]: col-xs-6 => col-6

**Comment by agriffard** [comment]
If I check 'Render title — If checked, the title will be rendered with the widget.', the title doesn't appear before the widget.

**Comment by sebastienros** [comment]
Title is rendered now

**Pull request merged**